vunit/test/unit/test_test_suites.py

```diff
@@ -105,5 +82,74 @@ def _read_test_results(contents, expected_test_cases):
                           config=None,
                           elaborate_only=False,
                           test_suite_name=None,
-                          test_cases=expected_test_cases)
-            return run._read_test_results(file_name=file_name)  # pylint: disable=protected-access
+                          test_cases=expected)
+            results = run._read_test_results(file_name=file_name)  # pylint: disable=protected-access
+            if do_assert:
+                self.assertEqual(results, expected)
+            return results
+
+    def test_exit_code(self):
+        """
+        Test that results are overwritten when none is FAILED but the exit code is nonzero
+        """
+
+        def test(contents, results, expected=None, werechecked=None):
+            """
+            Test the four combinations of 'sim_ok' and 'has_valid_exit_code'
+            """
+            if werechecked is None:
+                werechecked = [True, True, True, True]
+            self._test_exit_code(contents, results, True, False, werechecked[0])
+            self._test_exit_code(contents, results, False, False, werechecked[1])
+            self._test_exit_code(contents, results, True, True, werechecked[2])
+            val = results
+            if expected is not None:
+                val = expected
+            self._test_exit_code(contents, val, False, True, werechecked[3])
+
+        test(
+            """\ntest_start:test1\ntest_suite_done\n""",
+            {"test1": PASSED},
+            {"test1": FAILED},
+            [False, False, False, True]
+        )
+
+        test(
+            """\ntest_start:test1\ntest_suite_done\n""",
+            {"test1": PASSED, "test2": SKIPPED},
+            {"test1": FAILED, "test2": SKIPPED},
+            [False, False, False, True]
+        )
+
+        test("""\ntest_start:test1\n""", {"test1": FAILED, "test2": SKIPPED})
+        contents = """\ntest_start:test1\ntest_start:test2\n"""
+        test(contents, {"test1": PASSED, "test2": FAILED})
+        test(contents, {"test1": PASSED, "test2": FAILED, "test3": SKIPPED})
+
+    def _test_exit_code(self, contents, expected, sim_ok=True, has_valid_exit_code=False, waschecked=False):

```

Comment:
Terminology is not consistent, waschecked vs. done

Reply:
See comment above.